pandas/_libs/lib.pyx

```diff
@@ -1089,6 +1100,79 @@ def is_float(obj: object) -> bool:
     return util.is_float_object(obj)
 
 
+cpdef bint is_hashable(object obj):
+    """
+    Return True if hash(obj) will succeed, False otherwise.
+
+    Some types will pass a test against collections.abc.Hashable but fail when
+    they are actually hashed with hash().
+
+    Distinguish between these and other types by trying the call to hash() and
+    seeing if they raise TypeError.
+
+    Returns
+    -------
+    bool
+
+    Examples
+    --------
+    >>> import collections
+    >>> from pandas.api.types import is_hashable
+    >>> a = ([],)
+    >>> isinstance(a, collections.abc.Hashable)
+    True
+    >>> is_hashable(a)
+    False
+    """
+    cdef:
+        bint is_none
+        bint is_long
+        bint is_float
+        bint is_unicode
+        bint is_tuple
+        bint is_frozen_set
+        bint is_dict
+        bint is_list
+        bint is_any_set
+
+    # Perform all checks in order to be nice to the branch predictor
+    is_none = obj is None
+    is_long = PyLong_CheckExact(obj)
+    is_float = PyFloat_CheckExact(obj)
+    is_unicode = PyUnicode_CheckExact(obj)
+    is_tuple = PyTuple_CheckExact(obj)
```

Comment:
Is there any particular reason to use `CheckExact` over `Check` for the container types? I think the latter is preferable, given subclasses of these aren't that uncommon

Reply:
The motivation for this is pandas' own `FrozenList` which is hashable and a subclass of `list`. I agree that `PyXXXX_Check(...)` would be more intuitive but due to corner cases like `FrozenList` I think we need the exact checks to maintain the correct behaviour